### PR TITLE
changing trapz to trapezoid

### PR DESCRIPTION
**Pull request by NivekT**
This PR resolves issue #52606 while also adding support for complex number 

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61616
* #61615
* **#61475**

Differential Revision: [D29802844](https://our.internmc.facebook.com/intern/diff/D29802844)

### CODE REVIEW

File: aten/src/ATen/native/Integration.cpp [hunk 1]
```diff
@@ -38,16 +39,17 @@ Tensor zeros_like_except(const Tensor& y, int64_t dim) {
...
+Tensor trapezoid(const Tensor& y, const Tensor& x, int64_t dim) {
     dim = maybe_wrap_dim(dim, y);
     // asking for the integral with zero samples is a bit nonsensical,
     // but we'll return "0" to match numpy behavior.
     if (y.size(dim) == 0) {
         return zeros_like_except(y, dim);
     }
+    TORCH_CHECK(y.scalar_type() != kBool && x.scalar_type() != kBool, "trapezoid: received a bool input for `x` or `y`, but bool is not supported")
     Tensor x_viewed;
     if (x.dim() == 1) {
```
Comment (mruberry) [inline aten/src/ATen/native/Integration.cpp:51]: I don't think this is what NumPy does, see

https://github.com/numpy/numpy/blob/3d83143c2c2f877ee921c4c7c6746d836c459839/numpy/lib/function_base.py#L4228

NumPy is very aggressive about reshaping `x`. For example:

```
y = np.array(((1, 2, 3), (4, 5, 6)))
x = np.array((1, 1))

np.trapz(y, x)
: array([0., 0.])

np.trapz(y, x, axis=0)
: array([0., 0., 0.])
```

This is because NumPy reshapes the results of `diff`ing `x` when `x` is one dimensional. 

I think it's OK to be more restrictive than NumPy as long as we're principled about it, but we should be clear about the deviation

Comment (mruberry) [reply aten/src/ATen/native/Integration.cpp:51]: Line 57 below let's `x_viewed` be `x` when `x` has zero dimensions or two or more dimensions. Is there a check that should be done on the size of `x` in these cases? In general, is the rule that if `x` is 1d it's interpreted as the spacing along the specified axis, and then that tensor is reshaped/broadcast to the shape of `y`, and if `x` is not 1d then `x` is simply broadcast/reshaped to the shape of `y`?

File: aten/src/ATen/native/Integration.cpp [hunk 2]
```diff
@@ -58,15 +60,27 @@ Tensor trapz(const Tensor& y, const Tensor& x, int64_t dim) {
...
 }
 
-Tensor trapz(const Tensor& y, double dx, int64_t dim) {
+Tensor trapezoid(const Tensor& y, const Scalar& dx, int64_t dim) {
     // see above
     if (y.size(dim) == 0) {
         return zeros_like_except(y, dim);
     }
-    return do_trapz(y, dx, dim);
+    TORCH_CHECK(y.scalar_type() != kBool, "trapezoid: received a bool input for `y`, but bool is not supported")
```
Comment (mruberry) [inline aten/src/ATen/native/Integration.cpp:71]: Nice checks

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -4281,9 +4281,13 @@
...
     CompositeExplicitAutograd: rot90
 
+- func: trapezoid.x(Tensor y, Tensor x, *, int dim=-1) -> Tensor
+
+- func: trapezoid.dx(Tensor y, *, Scalar dx=1, int dim=-1) -> Tensor
+
 - func: trapz.x(Tensor y, Tensor x, *, int dim=-1) -> Tensor
 
-- func: trapz.dx(Tensor y, *, float dx=1, int dim=-1) -> Tensor
+- func: trapz.dx(Tensor y, *, float dx=1.0, int dim=-1) -> Tensor
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:4290]: This might trigger a backwards compatibility check warning

File: test/test_autograd.py [hunk 4]
```diff
@@ -2789,15 +2789,14 @@ def test_cat_empty(self):
                               lambda a, b: torch.cat((a, b)),
                               True, f_args_variable, f_args_tensor, check_forward_ad=True)
 
-    def test_trapz(self):
+    def test_trapezoid(self):
```
Comment (mruberry) [inline test/test_autograd.py:2792]: Is this test redundant with the automatically generated tests in test_ops.py? Can it be removed?

File: torch/_torch_docs.py [hunk 5]
```diff
@@ -10730,12 +10730,165 @@ def merge_dicts(*dicts):
             [2, 2],
             [2, 3],
             [3, 3]])
+
+""")
+
+add_docstr(torch.trapezoid,
+           r"""
+trapezoid(y, x=None, *, dx=None, dim=-1) -> Tensor
```
Comment (mruberry) [inline torch/_torch_docs.py:10738]: I updated the related Quip document with a proposal for how to document this function

File: torch/_torch_docs.py [hunk 6]
```diff
@@ -10730,12 +10730,165 @@ def merge_dicts(*dicts):
...
+    >>> y = torch.tensor([[0, 10, 20], [0, 20, 40]])
+    >>> torch.trapezoid(y, dx=2.0, dim=0)
+    tensor([40., 30., 60.])
 """)
 
 add_docstr(torch.trapz,
            r"""
 trapz(y, x, *, dim=-1) -> Tensor
 
+This is deprecated, please use :func:`torch.trapezoid`.
```
Comment (mruberry) [inline torch/_torch_docs.py:10890]: Let's give this the same signature as trapezoid and just remove all documentation except:

"Alias for :func:\`torch.trapezoid.\`."

The documentation in _tensor_docs.py will also have to be updated with the correct signature, but it's OK for the method documentation to say it's the method version of `trapz`, it doesn't need to explain it's an alias.

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -7400,6 +7427,14 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                   dtypes=all_types_and_complex_and(torch.bool, torch.float16, torch.bfloat16),
                   supports_out=False,
                   sample_inputs_func=sample_repeat_tile),
+    OpInfo('trapz',
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:7430]: Would you add a TODO comment here that in the future `trapz` should be made a proper alias of `trapezoid`

**Review by mruberry — COMMENTED** [review]
Hey @NivekT! Looking good overall. I think this needs to be rebased and check out the docs Quip. Beyond the docs there are just a few inline comments I made for your review!

Looking forward to having trapezoid and these OpInfos and the updated docs in PyTorch!

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (NivekT) [reply aten/src/ATen/native/native_functions.yaml:4290]: I will un-do this.

File: torch/_torch_docs.py (hunk 5 above)
Comment (NivekT) [reply torch/_torch_docs.py:10738]: Okay, I will update this accordingly.

File: torch/_torch_docs.py (hunk 6 above)
Comment (NivekT) [reply torch/_torch_docs.py:10890]: Okay, I will change the documentation to "Alias for :func:`torch.trapezoid.`."

Let me know if you want me to add wording that indicate the function is "deprecated".

File: test/test_autograd.py (hunk 4 above)
Comment (NivekT) [reply test/test_autograd.py:2792]: Yes I believe it is redundant. Removed.

File: torch/testing/_internal/common_methods_invocations.py (hunk 7 above)
Comment (NivekT) [reply torch/testing/_internal/common_methods_invocations.py:7430]: There are two notable differences between `trapz` and `trapezoid`:
1. `dx` is Scalar for `trapezoid` and double for `trapz`
2. (This may not be different) We added `OpInfo` for both functions and they support complex tensor inputs

Given 1., can `trapz` still be an alias for `trapezoid`?

Comment (NivekT) [reply torch/testing/_internal/common_methods_invocations.py:7430]: I added the TODO comment for now nonetheless.

File: aten/src/ATen/native/Integration.cpp (hunk 1 above)
Comment (NivekT) [reply aten/src/ATen/native/Integration.cpp:51]: I will address these points separately:

1. Reshaping Differences (your first comment)
I agree that there is a difference in behavior here, but I think it only matters for the case where the specific `dim` of `x` has length 2 and the `dim` of `y` > 2. For example:

```
y = torch.tensor([0, 1, 2, 3])
x = torch.tensor([0, 1])
print(np.trapz(y.numpy(), x.numpy())) # Returns 4.5
print(torch.trapz(y, x)) # Fails
```
More importantly, this seems to be a conceptual different. The current implementation in torch is broadcasting `x` and numpy is broadcasting `dx` as you described. I would guess it was unintended?

**Let me know if I should make a code change to fix this. This may lead to backward compatibility issues with `trapz`, or `trapezoid` will start behaving differently from `trapz` for this case.**

2. "In general, is the rule that if `x` is 1d it's interpreted as the spacing along the specified axis, and then that tensor is reshaped/broadcast to size of `y` along the dim in question"

Yes, this is true and I think this behavior is fine (except for the case discussed in bullet 1 above).

3.  "and if `x` is not 1d then only the dimension dim of `x` is considered and it's also reshaped/broadcast to the size of `y` along the `dim` in question?"

No, if `x` is not 1d, all dimension of `x` is considered, not just the specified `dim`. For example, this fails for both numpy and torch:

```
y = torch.tensor([[0, 1, 2], [3, 4, 5], [6, 7, 8]])
x = torch.tensor([[0, 1, 2], [3, 4, 5]])
print(np.trapz(y.numpy(), x.numpy())) # Fails
print(torch.trapz(y, x)) # Fails
```
This fails in torch because this happens in the `trapezoid` function, the call to `.slice()` only decrease the dimension of `x` along `dim` by 1, but the other dimension values are untouched.
```
// Input x with dimension (2x3)
Tensor x_left = x_viewed.slice(dim, 0, -1); // becomes (2x2)
Tensor x_right = x_viewed.slice(dim, 1); // becomes (2x2)
Tensor dx = x_right - x_left; // (2x2)
```
Later on, when `dx` is passed into `do_trapezoid`:
```
// Input y with dimension (3x3)
Tensor left = y.slice(dim, 0, -1); // becomes (3x2)
Tensor right = y.slice(dim, 1); // becomes (3x2)
return ((left + right) * dx).sum(dim) / 2.; // throws an error because dimensions of dx (2x2) and left (3x2) don't match and a broadcast isn't possible
```
In summary, all the broadcasting takes place in `(left + right) * dx` of `do_trapezoid`. An error returns if the dimensions do not match during `(left + right) * dx`.

`x_view` within `trapezoid` merely turns a 1D `x` with dimension `(n,)` into dimension `(1,1,...,n,...,1,1)` based on `dim` and `y.dim()` so that `x` can be broadcasted at a later time.

Comment (NivekT) [reply aten/src/ATen/native/Integration.cpp:51]: Sorry for the long block of text, in the meantime, I am leaving the code unchanged.

Let me know if you have any suggestion.

Comment (mruberry) [reply aten/src/ATen/native/Integration.cpp:51]: Thanks for the analysis. Adding a comment explaining the behavior and citing the (slight) discrepancy with NumPy would be nice, too

**Review by mruberry — APPROVED** [review]
Nice work, @NivekT!

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29802844).

File: aten/src/ATen/native/Integration.cpp [hunk 8]
```diff
@@ -38,35 +40,53 @@ Tensor zeros_like_except(const Tensor& y, int64_t dim) {
...
+    TORCH_CHECK(!(dx.isComplex() or dx.isBoolean()), "trapezoid: Currently, we only support dx as a real number.");
```
Comment (malfet) [inline aten/src/ATen/native/Integration.cpp:80]: This is not a valid C++ is it?

Comment (NivekT) [reply aten/src/ATen/native/Integration.cpp:80]: You are right. Thanks for spotting and fixing it.
